### ISSUE TITLE
Requests come as GET (.*), what could be causing it?

### ISSUE THREAD

**Issue by JCMais** [elastic/apm-agent-nodejs#742]
**Describe the bug**
I'm using version `2.0.4` of the agent with Koa, everything works ok, but the requests transactions are being sent as `GET (.*)`. What would cause them to be called with that name?

Everything seems correct, just the name that seems wrong

**To Reproduce**
1. I'm using a simple koa app with a route `router.get('/healthcheck' ctx => { ctx.body = 'ok' })`

**Expected behavior**
It should send the transaction with the name of the route, `/healthcheck`.

**Environment (please complete the following information)**
- OS: Linux on GKE
- Node.js version: 10
- APM Server version: 6.5.2
- Agent version: 2.0.4

**How are you starting the agent? (please tick one of the boxes)**
- [x] Calling `agent.start()` directly (e.g. `require('elastic-apm-node').start(...)`)
- [ ] Requiring `elastic-apm-node/start` from within the source code
- [ ] Starting node with `-r elastic-apm-node/start`

I've enabled debug logs on the apm agent:
![image](https://user-images.githubusercontent.com/1041647/49978567-4c65fc80-ff32-11e8-996d-fbbcf71d8dfb.png)

**Comment by watson** [comment]
Would it be possible for you to share your route setup with us?

**Comment by JCMais** [comment]
@watson alight, I found a consistent way to reproduce this. We just need to add a middleware to the router using `.use`, and access a route that was defined before the middleware was added. Basically the following code will reproduce this issue:
```
  const app = new Koa()
  const router = new Router()

  router.get('/users/:id', async ctx => {
    ctx.body = { id: ctx.params.id }
  })

  router.use((ctx, next) => next())

  router.get('/users/:id/profile', async ctx => {
    ctx.body = { id: ctx.params.id }
  })

  app.use(router.routes())
  app.use(router.allowedMethods())

  return app
```

Making `GET /users/123` leaves the transaction name as `GET (.*)`, tho when requesting `GET /users/123/profile`, the transaction is correctly named `/users/:id/profile`.

---

This happens because, by default, `koa-router` add those middlewares with path `(.*)` so they catch every request:
https://github.com/alexmingoia/koa-router/blob/e0ae639/lib/router.js#L243-L281

If that is the case, `matched` here: https://github.com/elastic/apm-agent-nodejs/blob/30b46e3fa550ba20993234e677a53217afe1632f/lib/instrumentation/modules/koa-router.js#L17

Will look like this:
```
{ path:
   [ Layer {
       opts:
        { end: true,
          name: null,
          sensitive: false,
          strict: false,
          prefix: '',
          ignoreCaptures: undefined },
       name: null,
       methods: [ 'HEAD', 'GET' ],
       paramNames:
        [ { name: 'id',
            prefix: '/',
            delimiter: '/',
            optional: false,
            repeat: false,
            partial: false,
            asterisk: false,
            pattern: '[^\\/]+?' } ],
       stack: [ [AsyncFunction] ],
       path: '/users/:id',
       regexp:
        { /^\/users\/((?:[^\/]+?))(?:\/(?=$))?$/i
          keys:
           [ { name: 'id',
               prefix: '/',
               delimiter: '/',
               optional: false,
               repeat: false,
               partial: false,
               asterisk: false,
               pattern: '[^\\/]+?' } ] } },
     Layer {
       opts:
        { end: false,
          name: undefined,
          sensitive: false,
          strict: false,
          prefix: '',
          ignoreCaptures: true },
       name: null,
       methods: [],
       paramNames:
        [ { name: 0,
            prefix: '',
            delimiter: '/',
            optional: false,
            repeat: false,
            partial: false,
            asterisk: false,
            pattern: '.*' } ],
       stack: [ [Function] ],
       path: '(.*)',
       regexp:
        { /^((?:.*))(?:\/(?=$))?(?=\/|$)/i
          keys:
           [ { name: 0,
               prefix: '',
               delimiter: '/',
               optional: false,
               repeat: false,
               partial: false,
               asterisk: false,
               pattern: '.*' } ] } } ],
  pathAndMethod:
   [ Layer {
       opts:
        { end: true,
          name: null,
          sensitive: false,
          strict: false,
          prefix: '',
          ignoreCaptures: undefined },
       name: null,
       methods: [ 'HEAD', 'GET' ],
       paramNames:
        [ { name: 'id',
            prefix: '/',
            delimiter: '/',
            optional: false,
            repeat: false,
            partial: false,
            asterisk: false,
            pattern: '[^\\/]+?' } ],
       stack: [ [AsyncFunction] ],
       path: '/users/:id',
       regexp:
        { /^\/users\/((?:[^\/]+?))(?:\/(?=$))?$/i
          keys:
           [ { name: 'id',
               prefix: '/',
               delimiter: '/',
               optional: false,
               repeat: false,
               partial: false,
               asterisk: false,
               pattern: '[^\\/]+?' } ] } },
     Layer {
       opts:
        { end: false,
          name: undefined,
          sensitive: false,
          strict: false,
          prefix: '',
          ignoreCaptures: true },
       name: null,
       methods: [],
       paramNames:
        [ { name: 0,
            prefix: '',
            delimiter: '/',
            optional: false,
            repeat: false,
            partial: false,
            asterisk: false,
            pattern: '.*' } ],
       stack: [ [Function] ],
       path: '(.*)',
       regexp:
        { /^((?:.*))(?:\/(?=$))?(?=\/|$)/i
          keys:
           [ { name: 0,
               prefix: '',
               delimiter: '/',
               optional: false,
               repeat: false,
               partial: false,
               asterisk: false,
               pattern: '.*' } ] } } ],
  route: true }
```

As we can see, `pathAndMethod` has two elements, the first one is the route matched, and ended the request, second one is the middleware we added to the router. 

The current instrumentation uses the last item to retrieve the route name:
https://github.com/elastic/apm-agent-nodejs/blob/master/lib/instrumentation/modules/koa-router.js#L25

So it gets named `(.*)`.

I think the current behavior is not correct, the first route added above does not wait for `next`, so the request ends there, and following routes/middlewares are never called.

However I don't see an easy way to fix this. One possibility is to check the list of arguments of each route stack item, and if one does not have length 2, it's a route without `next`, like the example above. But idk if that is reliable.

**Comment by watson** [comment]
Good catch! And thanks for sharing the code so we can reproduce this.

This sounds very similar to the issues we had with Express a long time ago, and we found a way to fix it there - so I'm pretty confident we can probably fix this bug as well.

**Comment by JCMais** [comment]
Btw, here is a patch for what I mentioned above, checking argument list of the middlewares:
```patch
patch-package
--- a/node_modules/elastic-apm-node/lib/instrumentation/modules/koa-router.js
+++ b/node_modules/elastic-apm-node/lib/instrumentation/modules/koa-router.js
@@ -22,7 +22,20 @@ module.exports = function (Router, agent, version, enabled) {
       }
 
       if (matched && matched.pathAndMethod && matched.pathAndMethod.length) {
-        var match = matched.pathAndMethod[matched.pathAndMethod.length - 1]
+        var match =
+          matched.pathAndMethod.reduce((matchedLayer, layer) => {
+            if (matchedLayer) return matchedLayer
+
+            const hasMiddlewareWithoutNext = layer.stack.reduce(
+              (hasFinalMiddleware, middleware) => {
+                if (hasFinalMiddleware) return hasFinalMiddleware
+                return middleware.length <= 1
+              },
+              false,
+            )
+
+            return hasMiddlewareWithoutNext ? layer : null
+          }, null) || matched.pathAndMethod[matched.pathAndMethod.length - 1]
         var path = match && match.path
         if (typeof path === 'string') {
           var name = method + ' ' + path
```

It fixes the issue for me, seems kinda hacky, but I'm pretty sure it's safe to assume that if a middleware has no `next` argument, the request is going to end there. Unless I'm missing something.

**Comment by watson** [comment]
While researching this bug, I found a few other places when using koa-router where transactions might get the wrong name. This have now all been fixed in master and will be part of the next release 😃

**Comment by JCMais** [comment]
Thanks @watson , will test as soon it's released

**Comment by watson** [comment]
Fixed as part of v2.0.6
